_posts/2025-02-13-rfprs.md

```diff
@@ -0,0 +1,372 @@
+convinced me that his "Reviewer-First" methodology was worth adopting, and I've been promoting it
+ever since.
+
+> At the heart of RFPRs is the concept that PRs, like a book, should tell a story.
+
+When reading a book, you typically know what it's about before you start because you've read the
+intro on the back. You start at the first chapter, and continue reading until the end. Even if the
+book was particularly long or complex, the chapters were structured in a way to guide you through
+the story that the author wanted to tell.
+
+RFPRs operate under the same principle. A reviewer should be able to open the PR and read the
+description to get a good idea for what the PR is addressing and what to expect inside. RFPRs are
+structured so that they can be reviewed commit-by-commit, in order, leaving feedback as they go.
+
+> Large or complex PRs are made more approachable if the author guides the reviewers.
+
+# RFPR Guidelines
+
+Following are guidelines for creating RFPRs. Afterwards we'll discuss some tooling that will make
+following the guidelines much easier.
+
+## PR titles
+
+PR titles should describe what the PR accomplishes.
+
+Examples of good PR titles:
+
+- [CONTROL_42] Expose gRPC endpoints for all controlplane APIs.
+- kafka: Create new topic for user events and enable mirrormaker for all regions.
+- Modify CMS domain models to use composition instead of inheritance.
+
+PR titles which should be avoided:
+
+- [CONTROL_42] (just a JIRA ticket number)
+- fix bug
+- (nothing)
+
+Imagine that you are trying to find a particular PR on GitHub during an incident. The PR title
+should be descriptive enough so that it can be easily identified.
+
+## PR descriptions
+
+The PR description sets the stage for the reviewer. Since they are going to review the PR
+commit-by-commit, the description should give them as much context as possible before they start.
+Assume that the reviewer does not have the context that you do.
+
+The description should roughly include these components:
+
+1. Links to tickets or other prior art.
+2. Overview of the problem or business goal.
+3. Description of your approach.
+4. Callouts to any parts which are complex or about which you are hoping for extra feedback.
+5. Edge cases and risks.
+
+It's worth noting that PRs which are small in scope will have a correspondingly smaller PR
+description. PRs which are large or complex will have a larger PR description. This is largely
+subjective, and at first I think it might feel like it's unnecessary work, but that is temporary. In
+fact I've found that by writing a detailed description, I will sometimes remember things I had
+forgotten to address during development. By writing the description *for the reviewer* this causes
+you to think about the things that the reviewer would probably have questions about. Anything you
+catch at this phase is a major win, as it is a short amount of time compared to the revisions to the
+PR which would otherwise follow.
```

Comment:
Would it be worth adding some mention of when it may be a better idea to split a large PR not by commits, but actually into separate PRs themselves? How would you define the demarcations for separate PRs and would it be based on granular deployment of code to a production environment?

Reply:
I feel like this is probably pretty difficult to talk about without adding a lot more content. I think you're right that a lot of it has to do with how a team does deployments, and whether or not you're incrementally deploying something. It seems related, but somewhat orthogonal (ten dollar word) to RFPRs.

Reply:
That's a good point. My experience with RFPRs has been that the process of building these commits makes it obvious when I have substantial unrelated work. At that point it's easy enough to move the commits to a new branch and PR.